halo2_proofs/src/arithmetic.rs

```diff
@@ -56,9 +60,35 @@ fn multiexp_serial<C: CurveAffine>(coeffs: &[C::Scalar], bases: &[C], acc: &mut
         tmp as usize
     }
 
-    let segments = (256 / c) + 1;
+    // Ideally `segments` should be calculated from the max number of bits among all scalars. But this requires a scan of all scalars, so we don't implement it for now.
+    let segments = (C::Scalar::NUM_BITS as usize + c - 1) / c;
+
+    // this can be optimized
+    let mut coeffs_in_segments = Vec::with_capacity(segments);
+    // track what is the last segment where we actually have nonzero bits, so we completely skip buckets where the scalar bits for all coeffs are 0
+    let mut max_nonzero_segment = None;
+    for current_segment in 0..segments {
+        let coeff_segments: Vec<_> = coeffs
+            .iter()
+            .map(|coeff| {
+                let c_bits = get_at::<C::Scalar>(current_segment, c, coeff);
+                if c_bits != 0 {
+                    max_nonzero_segment = Some(current_segment);
+                }
+                c_bits
+            })
+            .collect();
+        coeffs_in_segments.push(coeff_segments);
+    }
 
-    for current_segment in (0..segments).rev() {
+    if max_nonzero_segment.is_none() {
+        return;
+    }
+    for coeffs_seg in coeffs_in_segments
+        .into_iter()
+        .take(max_nonzero_segment.unwrap() + 1)
+        .rev()
+    {
         for _ in 0..c {
             *acc = acc.double();
```

Comment:
I think it would be faster and easier to maintain to add a boolean guard `if is_init` here
that defaults to false and is set to true on first initialization of the accumulator.

It would avoid an extra pass over the data.

Reply:
Ah yes, nice suggestion.

But now I am tempted to just do a full scan ahead of time to do this bit column optimization...